client/core/bookie.go

```diff
@@ -26,29 +27,78 @@ var (
 	outdatedClientErr = errors.New("outdated client")
 )
 
-// BookFeed manages a channel for receiving order book updates. The only
-// exported field, C, is a channel on which to receive the updates as a series
-// of *BookUpdate. It is imperative that the feeder (*BookFeed).Close() when no
-// longer using the feed.
-type BookFeed struct {
-	C     chan *BookUpdate
-	id    uint32
-	close func(*BookFeed)
+// BookFeed manages a channel for receiving order book updates. It is imperative
+// that the feeder (BookFeed).Close() when no longer using the feed.
+type BookFeed interface {
+	Next() <-chan *BookUpdate
+	Close()
+	Candles(dur string) error
 }
 
-// NewBookFeed is a constructor for a *BookFeed. The caller must Close() the
-// feed when it's no longer being used.
-func NewBookFeed(close func(feed *BookFeed)) *BookFeed {
-	return &BookFeed{
-		C:     make(chan *BookUpdate, 256),
-		id:    atomic.AddUint32(&feederID, 1),
-		close: close,
-	}
+// bookFeed implements BookFeed.
+type bookFeed struct {
+	// c is the update channel. Access to c is synchronized by the bookie's
+	// feedMtx.
+	c      chan *BookUpdate
+	bookie *bookie
+	id     uint32
+}
+
+// Next returns the channel for receiving updates.
+func (f *bookFeed) Next() <-chan *BookUpdate {
+	return f.c
 }
```

Comment:
Is it expected that the bookies feedMtx be held when calling this?

Reply:
Just for the sends and close, I believe.

Reply:
On this topic, the safest pattern with channels imo is to have the same goroutine that sends on the channel also be responsible for closing it.  Doing that means having a dedicated send loop and when the loop breaks it closes the channel to signal to receivers.

Not that I have a clear picture of how the bookie stuff would ideally look, but these pieces have always been pretty difficult to reason about.